File: tools/decompress_szdd.py
#!/usr/bin/env python3
"""Decompress SZDD (Microsoft compress) files"""
import struct
import sys

def decompress_szdd(input_path, output_path):
    with open(input_path, 'rb') as f:
        # Read header
        magic = f.read(8)
        if magic[:4] != b'SZDD':
            print(f"Not an SZDD file: {magic[:4]}")
            return False

        # Get compression mode and missing char
        comp_mode = magic[4]
        missing_char = magic[5]

        # Read uncompressed size (4 bytes, little-endian)
        uncompressed_size = struct.unpack('<I', f.read(4))[0]
        print(f"Uncompressed size: {uncompressed_size}")

        # Read compressed data
        compressed = f.read()

    # Decompress using LZ77 variant
    output = bytearray()
    ring_buffer = bytearray(4096)
    ring_pos = 4096 - 16

    i = 0
    while i < len(compressed) and len(output) < uncompressed_size:
        flags = compressed[i]
        i += 1

        for bit in range(8):
            if i >= len(compressed) or len(output) >= uncompressed_size:
                break

            if flags & (1 << bit):
                # Literal byte
                output.append(compressed[i])
                ring_buffer[ring_pos] = compressed[i]
                ring_pos = (ring_pos + 1) & 0xFFF
                i += 1
            else:
                # Back reference
                if i + 1 >= len(compressed):
                    break
                b1 = compressed[i]
                b2 = compressed[i + 1]
                i += 2

                offset = b1 | ((b2 & 0xF0) << 4)
                length = (b2 & 0x0F) + 3

                for _ in range(length):
                    if len(output) >= uncompressed_size:
                        break
                    byte = ring_buffer[offset]
                    output.append(byte)
                    ring_buffer[ring_pos] = byte
                    ring_pos = (ring_pos + 1) & 0xFFF
                    offset = (offset + 1) & 0xFFF

    with open(output_path, 'wb') as f:
        f.write(output)

    print(f"Decompressed {len(output)} bytes to {output_path}")
    return True

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print("Usage: decompress_szdd.py input.DA_ output.DTA")
        sys.exit(1)

    decompress_szdd(sys.argv[1], sys.argv[2])

File: tools/DecompressSzdd.cs
using System;
using System.IO;

class Program
{
    static void Main(string[] args)
    {
        if (args.Length != 2)
        {
            Console.WriteLine("Usage: DecompressSzdd input.DA_ output.DTA");
            return;
        }

        DecompressSzdd(args[0], args[1]);
    }

    static void DecompressSzdd(string inputPath, string outputPath)
    {
        using var fs = File.OpenRead(inputPath);
        using var reader = new BinaryReader(fs);

        // Read magic "SZDD"
        var magic = reader.ReadBytes(4);
        if (magic[0] != 'S' || magic[1] != 'Z' || magic[2] != 'D' || magic[3] != 'D')
        {
            Console.WriteLine($"Not an SZDD file");
            return;
        }

        // Read header
        var compMode = reader.ReadByte();
        var missingChar = reader.ReadByte();
        reader.ReadBytes(2); // padding

        var uncompressedSize = reader.ReadUInt32();
        Console.WriteLine($"Uncompressed size: {uncompressedSize}");

        // Read compressed data
        var compressed = reader.ReadBytes((int)(fs.Length - fs.Position));

        // Decompress
        var output = new MemoryStream();
        var ringBuffer = new byte[4096];
        int ringPos = 4096 - 16;

        int i = 0;
        while (i < compressed.Length && output.Length < uncompressedSize)
        {
            byte flags = compressed[i++];

            for (int bit = 0; bit < 8 && i < compressed.Length && output.Length < uncompressedSize; bit++)
            {
                if ((flags & (1 << bit)) != 0)
                {
                    // Literal byte
                    byte b = compressed[i++];
                    output.WriteByte(b);
                    ringBuffer[ringPos] = b;
                    ringPos = (ringPos + 1) & 0xFFF;
                }
                else
                {
                    // Back reference
                    if (i + 1 >= compressed.Length) break;

                    int b1 = compressed[i++];
                    int b2 = compressed[i++];

                    int offset = b1 | ((b2 & 0xF0) << 4);
                    int length = (b2 & 0x0F) + 3;

                    for (int j = 0; j < length && output.Length < uncompressedSize; j++)
                    {
                        byte b = ringBuffer[offset];
                        output.WriteByte(b);
                        ringBuffer[ringPos] = b;
                        ringPos = (ringPos + 1) & 0xFFF;
                        offset = (offset + 1) & 0xFFF;
                    }
                }
            }
        }

        File.WriteAllBytes(outputPath, output.ToArray());
        Console.WriteLine($"Decompressed {output.Length} bytes to {outputPath}");
    }
}
